### ISSUE TITLE
Phar class not found

### ISSUE THREAD

**Issue by SpiGAndromeda** [box-project/box#988]
## Bug report

| Question    | Answer
| ------------| ---------------
| Box version | 4.3.8@5534406
| PHP version | 8.2.4
| Platform with version | popos 22.04

The box compile command results in a `Class "Phar" not found in/.../autoload_runtime.php:11` error. PHAR is installed. Readonly is Off. The `'/usr/local/bin/composer' 'dump-autoload' '--classmap-authoritative' '--no-dev' '--ansi'` command doesn't fail if executed manually in the project directory.

<details>
 <summary>box.json.dist</summary>
 
 ```json
{
  "files-bin": [
    ".env.local.php",
    "autoload_runtime.template",
    "vendor/humbug/php-scoper/vendor-hotfix/.gitkeep"
  ],
  "directories": [
    "config",
    "public",
    "var"
  ],
  "force-autodiscovery": true,
  "check-requirements": false,
  "exclude-composer-files": false,
  "main": "bin/console"
}
 ```
</details>

<details>
 <summary>Output</summary>
 
 ```bash
 $ php vendor/bin/box compile
 > 
    ____
   / __ )____  _  __
  / __  / __ \| |/_/
 / /_/ / /_/ />  <
/_____/\____/_/|_|


Box version 4.3.8@5534406

 // Loading the configuration file                                                                                      
 // "/home/project-dir/box.json".                               

🔨  Building the PHAR "/home/project-dir/bin/console.phar"

? Checking Composer compatibility
    > Supported version detected
? No compactor to register
? Adding main file: /home/project-dir/bin/console
? Skip requirements checker
? Adding binary files
    > 3 file(s)
? Auto-discover files? Yes
? Exclude dev files? Yes
? Adding files
    > 5050 file(s)
? Generating new stub
  - Using shebang line: #!/usr/bin/env php
  - Using banner:
    > Generated by Humbug Box 4.3.8@5534406.
    > 
    > @link https://github.com/humbug/box
? Dumping the Composer autoloader

In ComposerOrchestrator.php line 207:
                                  
  Could not dump the autoloader.  
                                  

In ComposerOrchestrator.php line 210:
                                                                                                                    
  The command "'/usr/local/bin/composer' 'dump-autoload' '--classmap-authoritative' '--no-dev' '--ansi'" failed.    
                                                                                                                    
  Exit Code: 255(Unknown error)                                                                                     
                                                                                                                    
  Working directory: /tmp/box/Box37232                                                                              
                                                                                                                    
  Output:                                                                                                           
  ================                                                                                                  
  Generating optimized autoload files (authoritative)                                                       
                                                                                                                    
                                                                                                                    
  Error Output:                                                                                                     
  ================                                                                                                  
  Executing script cache:clear [KO]                                                                   
   [KO]                                                                                               
  Script cache:clear returned with error code 255                                                     
  !!  PHP Fatal error:  Uncaught Error: Class "Phar" not found in /tmp/box/Box37232/vendor/autoload_runtime.php:11  
  !!  Stack trace:                                                                                                  
  !!  #0 /tmp/box/Box37232/bin/console(10): require_once()                                                          
  !!  #1 {main}                                                                                                     
  !!    thrown in /tmp/box/Box37232/vendor/autoload_runtime.php on line 11                                          
  !!                                                                                                                
  Script @auto-scripts was called via post-autoload-dump
 ```
</details>

Because it's a Symfony project:

<details>
 <summary>autoload_runtime.template</summary>
 
 ```php
<?php

// autoload_runtime.php @generated by Symfony Runtime

$scriptFileName = $_SERVER['APP_SCRIPT_FILENAME'] ?? $_SERVER['SCRIPT_FILENAME'] ?? null;

if (true === (require_once __DIR__.'/autoload.php') || null === $scriptFileName || '' === $scriptFileName) {
    return;
}

$pharPath = Phar::running();

if (strlen($pharPath) > 0 && !str_starts_with($scriptFileName, 'phar://')) {
    $scriptFileName = $pharPath.DIRECTORY_SEPARATOR.$scriptFileName;
}

$app = require $scriptFileName;

if (!is_object($app)) {
    throw new TypeError(sprintf('Invalid return value: callable object expected, "%s" returned from "%s".', get_debug_type($app), $_SERVER['SCRIPT_FILENAME']));
}

$runtime = $_SERVER['APP_RUNTIME'] ?? $_ENV['APP_RUNTIME'] ?? %runtime_class%;
$runtime = new $runtime(($_SERVER['APP_RUNTIME_OPTIONS'] ?? $_ENV['APP_RUNTIME_OPTIONS'] ?? []) + %runtime_options%);

[$app, $args] = $runtime
    ->getResolver($app)
    ->resolve();

$app = $app(...$args);

exit(
    $runtime
        ->getRunner($app)
        ->run()
);

 ```
</details>

**Comment by theofidry** [comment]
This looks really strange: I really don't think it's possible for the PHAR class to not exist yet especially since Box itself uses it. Is the script used really the one you linked? Isn't there a namespace declared or something?

**Comment by SpiGAndromeda** [comment]
I can't explain it either. I tried executing box from global installation and from project installation. Same result.

I just added a `use Phar;` statement to the autoload_runtime template. The result is the same. Error line changed naturally. Which also means this template is used for sure.

I also checked if the Phar extension is enabled. It is.

**Comment by theofidry** [comment]
Would you have a reproducer by any chance?

Out of curiosity could you also try to run the Box Symfony E2E test?

**Comment by theofidry** [comment]
It should be make e2e_symfony_runtime or something

**Comment by SpiGAndromeda** [comment]
> Would you have a reproducer by any chance?
> 
> Out of curiosity could you also try to run the Box Symfony E2E test?

I created a minimal repository with which I was able to reproduce the issue: https://github.com/it-bens/box-test

**Comment by SpiGAndromeda** [comment]
The E2E test leads to the same result:

```bash
$ make e2e_symfony_runtime
> php fixtures/build/dir018/new-bin/console --version --no-ansi \
        1>dist/dir018/expected-stdout \
        2>dist/dir018/expected-stderr
fixtures/build/dir018/new-bin/console cache:pool:clear cache.global_clearer --env=prod --ansi

 // Calling cache clearer: cache.global_clearer                                                                         

                                                                                                                        
 [OK] Cache was successfully cleared.                                                                                   
                                                                                                                        

fixtures/build/dir018/new-bin/console cache:clear --env=prod --ansi

 // Clearing the cache for the prod environment with debug false                                                        

                                                                                                                        
 [OK] Cache for the "prod" environment (debug=false) was successfully cleared.                                          
                                                                                                                        

rm -rf fixtures/build/dir018/var/cache/prod/*
bin/box.phar compile --working-dir=fixtures/build/dir018 --no-parallel --ansi

    ____
   / __ )____  _  __
  / __  / __ \| |/_/
 / /_/ / /_/ />  <
/_____/\____/_/|_|


Box version dev-main@25cb301 2023-04-10 13:25:29 UTC

 // Loading the configuration file "/home/box/fixtures/build/dir018/box.json.dist".              

🔨  Building the PHAR "/home/box/fixtures/build/dir018/new-bin/console.phar"

? Checking Composer compatibility
    > Supported version detected
? No compactor to register
? Adding main file: /home/box/fixtures/build/dir018/new-bin/console
? Skip requirements checker
? Adding binary files
    > 3 file(s)
? Auto-discover files? Yes
? Exclude dev files? Yes
? Adding files
    > 1275 file(s)
? Generating new stub
  - Using shebang line: #!/usr/bin/env php
  - Using banner:
    > Generated by Humbug Box dev-main@25cb301.
    > 
    > @link https://github.com/humbug/box
? Dumping the Composer autoloader

In ComposerOrchestrator.php line 102:
                                  
  Could not dump the autoloader.  
                                  

In ComposerOrchestrator.php line 102:
                                                                                                                    
  The command "'/usr/local/bin/composer' 'dump-autoload' '--classmap-authoritative' '--no-dev' '--ansi'" failed.    
                                                                                                                    
  Exit Code: 255(Unknown error)                                                                                     
                                                                                                                    
  Working directory: /tmp/box/Box33769                                                                              
                                                                                                                    
  Output:                                                                                                           
  ================                                                                                                  
  Generating optimized autoload files (authoritative)                                                       
                                                                                                                    
                                                                                                                    
  Error Output:                                                                                                     
  ================                                                                                                  
  Executing script cache:clear [KO]                                                                   
   [KO]                                                                                               
  Script cache:clear returned with error code 255                                                     
  !!  PHP Fatal error:  Uncaught Error: Class "Phar" not found in /tmp/box/Box33769/vendor/autoload_runtime.php:11  
  !!  Stack trace:                                                                                                  
  !!  #0 /tmp/box/Box33769/new-bin/console(12): require_once()                                                      
  !!  #1 {main}                                                                                                     
  !!    thrown in /tmp/box/Box33769/vendor/autoload_runtime.php on line 11                                          
  !!                                                                                                                
  Script @auto-scripts was called via post-autoload-dump                                              
                                                                                                                    

compile [--debug] [--no-parallel] [--no-restart] [--dev] [--no-config] [--with-docker] [--composer-bin COMPOSER-BIN] [--allow-composer-check-failure] [-c|--config CONFIG] [-d|--working-dir WORKING-DIR]

make: *** [Makefile.e2e:318: e2e_symfony_runtime] Error 1
```

**Comment by theofidry** [comment]
> It looks like that some kind of minimal PHP version is used todo the dump. 

Is there such a thing available in your CI? Running the command with `-vvv` should be able to show it. But IIRC box doesn't try to pick another PHP version, only the Composer binary to dump the autoloader.

**Comment by maartendekeizer** [comment]
Not that I'm aware of. When running `php -i` / `php -m` as seperated job Phar is listed.
Output when running with `-vvv` https://github.com/Markei/kopio/actions/runs/6478641608/job/17590749216

**Comment by theofidry** [comment]
@maartendekeizer isn't your issue a timeout rather than `Phar` not found though?

**Comment by maartendekeizer** [comment]
The timeout did not happen without `-vvv`, than the Phar class issue will come up. See https://github.com/Markei/kopio/actions/runs/6474508303/job/17579581668
How can I change the timeout in the case of `-vvv`

**Comment by theofidry** [comment]
@maartendekeizer I am a bit confused by your CI though: https://github.com/Markei/kopio/blob/main/.github/workflows/build-phar.yml

How is PHP installed there? It appears you pick whatever ubuntu latest has which IMO is a bad idea. I am not sure if it's the root of the problem, as I am not aware of a PHP binary that comes _without_ the PHAR extension, but looking at your workflow that's the only thing I can pick up.

**Comment by theofidry** [comment]
I actually missed https://github.com/box-project/box/issues/988#issuecomment-1501796287. Might be another lead I'll have to check ASAP. I'm focusing on finishing to fix the `diff` command for the next release and can check this afterwards (or before the release)

**Comment by maartendekeizer** [comment]
Upgraded to 4.4.0, issue still exists.

Add a dump of [php_ini_scanned_files()](https://www.php.net/manual/en/function.php-ini-scanned-files.php). See https://github.com/Markei/kopio/actions/runs/6525975372/job/17719227836

In the step **Cache warm up** multiple ini files are loaded, each ini file will activate an extension (for instance Phar).
In the step **Run box** no other ini files are loaded, only the php.ini.

**Comment by maartendekeizer** [comment]
Setting BOX_ALLOW_XDEBUG=1 will fix the problem which bring us maybe to the composer xdebug handler https://github.com/composer/xdebug-handler/blob/main/src/XdebugHandler.php

**Comment by theofidry** [comment]
Indeed if removing the restart fixes it it likely a bug there. @maartendekeizer do you have any local (potentially via docker) reproducer? I would like to avoid pulling in johnstevenson without one if I can help it

**Comment by maartendekeizer** [comment]
Sorry local under Windows and no xdebug. Unsure how to reproduce this is under Docker, but the linked repository can be used as an example.

I see that Box is using XdebugHandler too. So we have the situation that

**Comment by theofidry** [comment]
@maartendekeizer I am a bit perplexed as I can't reproduce your issue here: https://github.com/box-project/box/pull/1089

**Comment by theofidry** [comment]
Hm actually didn't check but could also be from the Flex plugin too...

**Comment by theofidry** [comment]
I managed to reproduce and isolate the issue in https://github.com/box-project/box/actions/runs/6592295374/job/17912645537#logs.

My findings:

- It is related to the `symfony/flex` plugin
- It happens as soon as the process in which the plugin is used is restarted
- This happens _only_ for the `symfony-cmd` scripts, the `script` and `php-script` are both working fine
- It is NOT related to the Symfony command used (which makes sense, the problem is about the ini settings of the PHP process used).

Looking at https://github.com/symfony/flex/blob/3e8d6c423092ca89dd8fb998ddb9d7a91139e52f/src/ScriptExecutor.php#L94 I have **no clue** why this would result in a PHP process with non-sensical ini values.

For the record: this is not a problem of Box restarting the process. I mean it is linked, but Composer is executed within it and the PHP scripts started by flex also work fine.

So I am a bit at a loss

**Comment by maartendekeizer** [comment]
I figured out https://github.com/symfony/flex/blob/3e8d6c423092ca89dd8fb998ddb9d7a91139e52f/src/ScriptExecutor.php#L120C9-L120C9 is a bad way to find out if the xdebughandler has restarted composer.

The code will check for a specific environment variabel, if it set it assumed it will need to use another ini file instead of the used php.ini (in the case this tmp ini is created in the no-xdebug-restart and is passed via PHPRC env to composer).

Cleaning the env will not trigger the bad flex code and will continue to use the right tmpIni.

Successfull run: https://github.com/maartendekeizer/box/actions/runs/6593003554/job/17914732375
Diff here: https://github.com/theofidry/box/pull/3/files

**Comment by theofidry** [comment]
That does the trick! Prob worth fixing on Flex side :o

**Comment by theofidry** [comment]
Closed by #1091, thank you a lot @maartendekeizer for figuring it out
